### PR TITLE
Automate release, workaround macOS CI errors

### PR DESCRIPTION
**Pull request by w-bonelli**
Includes 

- Add automatic release workflow as per #178
- Symlink gfortran dylibs to `/usr/local/lib` on macOS CI so PRMS binaries included in this repo can find them. The current installation in CI via `awvwgk/setup-fortran` puts libs in `/usr/local/opt/gcc/lib/gcc/{version}` by default
- fix `pydot` usage in `pywatershed/analysis/model_graph.py` causing CI errors

The workflow currently assumes the repo adopts the same branching scheme as FloPy, where development happens on `develop` and code is merged to `main` at release time, while currently the repo merges PRs directly to `main`

If we adopt the distinction between trunk and development branch this should be good to go and the workflow will be triggered when a branch matching semver format with leading "v" (e.g. `v0.1.3`) is pushed. If we want to stick to the existing scheme and not introduce a development branch, I can update the workflow to trigger on a different condition, e.g. when commits on `main` are tagged as release candidates with semver number.

Before merging I can tack on some release docs

### CODE REVIEW

**Comment by jmccreight** [comment]
Thank you, Wes!
I prefer to only have a main branch which we will snapshot to version branches.

File: pywatershed/analysis/model_graph.py
Comment (jmccreight) [inline pywatershed/analysis/model_graph.py:101]: strange, thanks for doing that.

File: RELEASING.md
Comment (jmccreight) [inline RELEASING.md:13]: should `version.py` be `version.txt`. Will that be added the first time this is run?

**Review by jmccreight — COMMENTED** [review]
When the version number is then merged to main, is there some mechanism that differentiates additional/other commits with that version.txt from the actual version/specific commit? What I've done before is add M.m.p+ into a file like version.txt to avoid that confusion.

Also, the workflow appears to preclude updating all but the most recent release because of this final step that merges to main. IE, if you released 2.1.1, you could not go back and use the workflow to hotfix 1.8.1. Or do I misunderstand that?

**Review by jmccreight — COMMENTED** [review]
When the version number is then merged to main, is there some mechanism that differentiates additional/other commits with that version.txt from the actual version/specific commit? What I've done before is add M.m.p+ into a file like version.txt to avoid that confusion.

Also, the workflow appears to preclude updating all but the most recent release because of this final step that merges to main. IE, if you released 2.1.1, you could not go back and use the workflow to hotfix 1.8.1. Or do I misunderstand that?

File: RELEASING.md
Comment (jmccreight) [reply RELEASING.md:13]: or should it be `pywatershed/version.py`? (ie in the package and not the root?)

Comment (w-bonelli) [reply RELEASING.md:13]: Yes just `pywatershed/version.py` currently, will fix. Alternatively we could introduce a `version.txt` in the root like flopy, in that case `ci/scripts/update_version.py` would need an update to modify both

File: .github/workflows/release.yaml
Comment (w-bonelli) [reply .github/workflows/release.yaml:1]: I didn't see this before adding `RELEASING.md`, sorry. Will move to `README.md` here

I added git-cliff just to match what flopy does. I could remove it. It works best if commits are [conventional](https://www.conventionalcommits.org/en/v1.0.0/#summary), but can be [configured](https://git-cliff.org/docs/configuration#filter_unconventional) to accept any format with `conventional_commits = false`. With the defaults

```
conventional_commits = true
filter_unconventional = true
```

I don't think you'd need to squash since merge commits would be filtered out.

**Comment by w-bonelli** [comment]
The last step assumes most releases are minor (back-compatible features) and drafts a PR setting minor number +=1 patch 0 but it doesn't merge, for hotfixes or major releases the PR could be closed. But maybe it is better to just remove this and keep placeholder values in version files in `main` like you say and only set them for releases?

**Comment by langevin-usgs** [comment]
Hey guys, I know I'm late to the party here, but it if there is any way to set this all up to follow the approach that we have for MODFLOW and FloPy, I think that would be best.  We have a versioning approach and release strategy with main, develop, and tags.  We do not typically go back and make hot fixes to previous releases; instead we are always moving forward.  If we want to rethink that for EC in general, then we should, but otherwise we should try and stick with our current recipe.  My 2 cents.

**Comment by jdhughes-usgs** [comment]
> Thank you, Wes! I prefer to only have a main branch which we will snapshot to version branches.

I don't think there should be version branches. The commit with the version will be tagged and the release will point to a specific commit. So it will always be possible to get the source for a specific release.

Version branches are also inconsistent with all of our other repositories.

File: .github/workflows/release.yaml
Comment (jdhughes-usgs) [reply .github/workflows/release.yaml:1]: I think the README.md is good. It being in `ci/` seems ok since this isn't something needed by anyone but us since we are the only ones that will make releases.

File: RELEASING.md
Comment (jdhughes-usgs) [inline RELEASING.md:1]: I don't think this should be in the root directory. Having it within `ci/` or in `github/` is probably better since this is only something that maintainers will be doing.

Also `RELEASE.md` is probably a better name.

**Comment by jmccreight** [comment]
Version branches are required for the workflow. They can then be deleted. 

Me, Wes, and Chris just decided to make everything consistent with flopy with one minor exception of the version in version.txt being x.y.z+ on the develop instead of the same or ahead of x.y.z on main.

**Comment by jmccreight** [comment]
@jdhughes-usgs i assume you're good with the above. LMK if otherwise

File: ci/scripts/README.md
Comment (jdhughes-usgs) [inline ci/scripts/README.md:1]: This should be consolidated into `RELEASE.md` in `.github/`. I don't see why this information needs to be in more than one place.

**Review by jdhughes-usgs — CHANGES_REQUESTED** [review]

**Comment by jdhughes-usgs** [comment]
> @jdhughes-usgs i assume you're good with the above. LMK if otherwise

I am ok with it being like flopy, mf6, etc.

The version branches would be deleted since they are not needed after a release is made.

**Comment by jdhughes-usgs** [comment]
> Version branches are required for the workflow. They can then be deleted.
> 
> Me, Wes, and Chris just decided to make everything consistent with flopy with one minor exception of the version in version.txt being x.y.z+ on the develop instead of the same or ahead of x.y.z on main.

x.y.z+ on develop is ok. To be most useful would be great if it could be x.y.z+`commit_hash` on develop. @w-bonelli is there is a way this could be done as part of CI? If we could get this to work we could propagate this to the other repos.

**Comment by jdhughes-usgs** [comment]
> > Version branches are required for the workflow. They can then be deleted.
> > Me, Wes, and Chris just decided to make everything consistent with flopy with one minor exception of the version in version.txt being x.y.z+ on the develop instead of the same or ahead of x.y.z on main.
> 
> x.y.z+ on develop is ok. To be most useful would be great if it could be x.y.z+`commit_hash` on develop. @w-bonelli is there is a way this could be done as part of CI? If we could get this to work we could propagate this to the other repos.

@w-bonelli But we don't have to hold up this PR to implement this, if possible. Maybe this becomes an issue here or in flopy if you think it is possible.

**Comment by jmccreight** [comment]
> > Version branches are required for the workflow. They can then be deleted.
> > Me, Wes, and Chris just decided to make everything consistent with flopy with one minor exception of the version in version.txt being x.y.z+ on the develop instead of the same or ahead of x.y.z on main.
> 
> x.y.z+ on develop is ok. To be most useful would be great if it could be x.y.z+`commit_hash` on develop. @w-bonelli is there is a way this could be done as part of CI? If we could get this to work we could propagate this to the other repos.

what is the `commit hash`, that of the current revision? it's hard/impossible to know the commit hash in advance of committing.

**Comment by jdhughes-usgs** [comment]
> > > Version branches are required for the workflow. They can then be deleted.
> > > Me, Wes, and Chris just decided to make everything consistent with flopy with one minor exception of the version in version.txt being x.y.z+ on the develop instead of the same or ahead of x.y.z on main.
> > 
> > 
> > x.y.z+ on develop is ok. To be most useful would be great if it could be x.y.z+`commit_hash` on develop. @w-bonelli is there is a way this could be done as part of CI? If we could get this to work we could propagate this to the other repos.
> 
> what is the `commit hash`, that of the current revision? it's hard/impossible to know the commit hash in advance of committing.

May not be possible but a google search found this https://github.com/marketplace/actions/commit-hash

**Comment by w-bonelli** [comment]
We have a [flopy issue on this topic](https://github.com/modflowpy/flopy/issues/1664). Two potential solutions are

- [versioneer](https://github.com/python-versioneer/python-versioneer)
- [setuptools_scm](https://github.com/pypa/setuptools_scm/)

I think https://github.com/marketplace/actions/commit-hash can extract version+hash for consumption by other tools (e.g. Docker) but cannot automatically embed a fine-grained version string in the module, which seems like what we want.

File: RELEASING.md
Comment (w-bonelli) [reply RELEASING.md]: moved to `.github/RELEASE.md`

File: ci/scripts/README.md
Comment (w-bonelli) [reply ci/scripts/README.md]: consolidated into `.github/RELEASE.md`

**Comment by w-bonelli** [comment]
I think this is ready. The procedure is described in `.github/RELEASE.md`. Here is a test run on my fork:

- [Initial run](https://github.com/w-bonelli/pywatershed/actions/runs/5057147916) triggered on `v0.1.3` branch push
- [Release PR](https://github.com/w-bonelli/pywatershed/pull/14) opened from release branch into `main`. The changelog only shows 1 commit because currently `git-cliff` is configured to filter out non-conventional commit messages, and omit CI-related commits &mdash; see [flopy releases](https://github.com/modflowpy/flopy/releases/tag/3.3.6) for an example of what this will look like in practice.
- [Run triggered](https://github.com/w-bonelli/pywatershed/actions/runs/5057974727) by publishing [the release](https://github.com/w-bonelli/pywatershed/releases/tag/0.1.3) drafted by the initial run. The publish step failed because my fork is not configured for trusted publishing, it should succeed on this repo
- [Reset PR](https://github.com/w-bonelli/pywatershed/pull/15) opened into `develop`

The publish step assumes a `pypi` [deployment environment](https://github.com/EC-USGS/pywatershed/settings/environments) exists and this repo is configured on PyPI for [trusted publishing](https://docs.pypi.org/trusted-publishers/using-a-publisher/#the-easy-way) (I did both). There is no need to store PyPI credentials as repo secrets.

**Comment by w-bonelli** [comment]
In the test run above, the workflow mistakenly leaves a "+" in `version.txt` in the released version, and increments the patch version number on the final reset PR, both of which were unwanted leftovers from the flopy approach. I just fixed the former and removed the latter step, so the workflow now follows the procedure as described in `.github/RELEASE.md` (no more incrementing the patch). Also bumped `pywatershed/version.py` back down to 0.1.2 since that is the most recent released version.

**Comment by jdhughes-usgs** [comment]
@w-bonelli are you satisfied with the PR now? If so I will merge unless @jmccreight objects.

**Review by jmccreight — APPROVED** [review]
Thanks alot for cleaning up the envs and restructuring the markdowns.

**Comment by jmccreight** [comment]
This PR almost closes #165.

**Comment by jmccreight** [comment]
Wes, this is awesome. Thanks for all your work on this.
I did your "what's new" notes for you, in case you were curious to see that. 
I'll merge this after all the tests pass again.
closes #178

**Review by jmccreight — APPROVED** [review]

**Comment by w-bonelli** [comment]
@jmccreight sure thing. Thanks, sorry, forgot to do the what's new!

**Pull request merged**
